jdbc-handler/src/main/java/org/apache/hive/storage/jdbc/DBRecordWritable.java

```diff
@@ -61,16 +60,30 @@ public void write(PreparedStatement statement) throws SQLException {
     if (columnValues == null) {
       throw new SQLException("No data available to be written");
     }
-    ParameterMetaData parameterMetaData = statement.getParameterMetaData();
     for (int i = 0; i < columnValues.length; i++) {
       Object value = columnValues[i];
-      if ((parameterMetaData.getParameterType(i + 1) == Types.CHAR) && value != null && value instanceof Boolean) {
+      if ((checkParamMeta(statement) && statement.getParameterMetaData().getParameterType(i + 1) == Types.CHAR)
+          && value != null && value instanceof Boolean) {
         value = ((Boolean) value).booleanValue() ? "1" : "0";
       }
       statement.setObject(i + 1, value);
```

Comment:
Here are doing a conversion based on expected and actual type. Instead of relying on parameterMetaData which may not be available in every JDBC driver can't we do the conversion earlier. E.g., when we setup `columnValues`?

Reply:
I have created `DerbyRecordWritable`which extends `DBRecordWritable`for derby, and other jdbc datasources(eg. mysql, pg) can use original `DBRecordWritable` which does not rely on parameterMetaData .  Please check again. Thank you.

Reply:
> can't we do the conversion earlier. E.g., when we setup columnValues?

I think not. type conversion depends on actual type which can only be retrieved from underly jdbc-datasource using jdbc api `parameterMetaData`.